**Домашняя работа**
1.   Определить "нетто-доходность" i-ой (i = 1, 2, 3) акции в день
2.   Вычислить выборочные средние и дисперсии доходности для каждой акции, выборочные ковариации
3.   Составить задачу минимизации дисперсии портфеля при заданной средней доходности портфеля
4.   Решить задачу аналитически, применяя правило Лагранжа, обосновать существование решения в задаче
5.   Найти решение численно, реализовав метод множителей Лагранжа на Питоне




In [2]:
import pandas as pd 
import numpy as np 
import sympy as sm

In [37]:
#Считываем данные 3-х компаний из файлов
first_action_AVAN = pd.read_excel('drive/MyDrive/мо/AVAN.xlsx')
second_action_MVID = pd.read_excel('drive/MyDrive/мо/MVID.xlsx')
third_action_VJGZ = pd.read_excel('drive/MyDrive/мо/VJGZ.xlsx')

In [4]:
#Список цен 3-х акций
r_avan = list(map(float, first_action_AVAN['Цена'].tolist()))
r_mvid = list(map(float, second_action_MVID['Цена'].tolist()))
r_vjgz =  list(map(float, third_action_VJGZ['Цена'].tolist()))

In [5]:
#Список из доходностей для 3-х компаний
r_avan_1 = [(r_avan[i + 1] / r_avan[i]) - 1 for i in range(200)]
r_mvid_2 = [(r_mvid[i + 1] / r_mvid[i]) - 1 for i in range(200)]
r_vjgz_3 = [(r_vjgz[i + 1] / r_vjgz[i]) - 1 for i in range(200)]

In [6]:
#Определим "нетто-доходность" i-ой акции в день t
t = int(input("Введите номер дня (от 1 до 200) за который хотите получить доходность: "))
i = int(input("Введите номер компании, где AVAN - 1, MVID - 2, VJGZ - 3: "))
while t < 1 or t > 200:
    t = int(input(f'Введите номер дня (от 1 до 200) еще раз: '))
else:
    if i == 1:
        print(f'Доходность акций компании AVAN за {t} день равна {r_avan_1[i - 1]}')
    elif i == 2:
        print(f'Доходность акций компании MVID за {t} день равна {r_mvid_2[i - 1]}')
    elif i == 3:
        print(f'Доходность акций компании VJGZ за {t} день равна {r_vjgz_3[i - 1]}')
    else:
        print('Введите правильные данные')

Введите номер дня (от 1 до 200) за который хотите получить доходность: 3
Введите номер компании, где AVAN - 1, MVID - 2, VJGZ - 3: 3
Доходность акций компании VJGZ за 3 день равна 0.010544815465729274


1**Воспользуюсь следующими формулами**

> Выборочной средней называется среднее арифметическое всех значений выборки: $M = 1/n \sum_{i = 0}^n r_i$


> Выборочная дисперсия - это среднее арифметическое квадратов отклонения всех вариантов выборки от ее средней: $D = 1/n \sum_{i = 0} ^ n (r_i - M)^2$




In [7]:
#Выборочные средние для каждой акции
M_avan_1 = sum(r_avan_1) / len(r_avan_1)
M_mvid_2 = sum(r_mvid_2) / len(r_mvid_2)
M_vjgz_3 = sum(r_vjgz_3) / len(r_vjgz_3)

print(f'Для акций AVAN выборочное среднее = {M_avan_1}\n'
f'Для акций MVID выборочное среднее = {M_mvid_2}\n'
f'Для акций VJGZ выборочное среднее = {M_vjgz_3}')

Для акций AVAN выборочное среднее = 0.0011247630175105704
Для акций MVID выборочное среднее = 0.006043591894223941
Для акций VJGZ выборочное среднее = 0.0031135893613838954


Выберу такое M*, чтобы выполнялось следующее условие: $min M_i < M* < max M_i$

In [8]:
#Выбор M*
all_M = [M_avan_1, M_mvid_2, M_vjgz_3]
M_range = float(input(f'Введите значение M* из диапазона'
f'({min(all_M),  max(all_M)}): '))
while M_range <= min(all_M) or M_range >= max(all_M):
    M_range = float(input(f'Введите значение из диапазона'
    f'({min(all_M),  max(all_M)}) еще раз: '))
else:
    print(f'Вы ввели M* = {M_range}')

Введите значение M* из диапазона((0.0011247630175105704, 0.006043591894223941)): 0.004
Вы ввели M* = 0.004


In [9]:
#Выборачная дисперсия для каждой акции
D_avan_1 = sum([(r_avan_1[i] - M_avan_1)**2 for i in range(len(r_avan_1))]) / len(r_avan_1)
D_mvid_2 = sum([(r_mvid_2[i] - M_mvid_2)**2 for i in range(len(r_mvid_2))]) / len(r_mvid_2)
D_vjgz_3 = sum([(r_vjgz_3[i] - M_vjgz_3)**2 for i in range(len(r_vjgz_3))]) / len(r_vjgz_3)

print(f'Для акций AVAN выборочная дисперсия в квадрате = {D_avan_1}\n'
f'Для акций MVID выборочная дисперсия в квадрате = {D_mvid_2}\n'
f'Для акций VJGZ выборочная дисперсия в квадрате = {D_vjgz_3}')

Для акций AVAN выборочная дисперсия в квадрате = 0.0014168584314022694
Для акций MVID выборочная дисперсия в квадрате = 0.0022635154092858446
Для акций VJGZ выборочная дисперсия в квадрате = 0.0025429379551753684


In [10]:
#Матрица выборочных ковариаций
data = np.array([r_avan_1, r_mvid_2, r_vjgz_3])
C = np.cov(data, bias = True)

print(C)

[[ 1.41685843e-03  8.63479402e-05 -2.49744668e-04]
 [ 8.63479402e-05  2.26351541e-03  1.59762813e-03]
 [-2.49744668e-04  1.59762813e-03  2.54293796e-03]]


**Интерпретация ковариационной матрицы**

> Значения по главной диагонали матрицы - это дисперсия каждой акции. Например:

1.   Для акций AVAN = 1.41685843e-03
2.   Для акций MVID = 2.26351541
3.   Для акций VJGZ = 2.54293796e-03


> Другие значения в матрице представляют собой ковариации между различными акциями. Например:


1.   Ковариация между акциями AVAN и MVID составляет 8.63479402e-05
2.   Ковариация между акциями AVAN и VJGZ составляет -2.49744668e-04
3.   Ковариация между акциями MVID и VJGZ составляет 1.59762813e-03



> *Положительное число* для ковариации указывает, что две акции имеют тенденцию увеличиваться или уменьшаться в тандеме.

> *Отрицательное число* ковариации указывет на то, что по мере увеличения одной акции вторая акция имеет тенденцию к уменьшению.
















In [11]:
# Запишу целевую функцию (Cx, x)
x1, x2, x3 = sm.symbols('x1 x2 x3')
X = np.array([x1, x2, x3])
Cx = C.dot(X)
Cx_x = np.dot(Cx, X)

print(f'(Cx, x) = {Cx_x} -> min')

(Cx, x) = x1*(0.00141685843140227*x1 + 8.63479402469543e-5*x2 - 0.000249744668073191*x3) + x2*(8.63479402469543e-5*x1 + 0.00226351540928585*x2 + 0.00159762813078828*x3) + x3*(-0.000249744668073191*x1 + 0.00159762813078828*x2 + 0.00254293795517537*x3) -> min


**Запишу ограничения в нашей задачи для трех акций**

> $\sum_{i = 1}^3 x_i - 1 ==> x_1 + x_2 + x_3 - 1$

> $\sum_{i = 1}^3 M_ix_i - M* ==> M_1x_1 + M_2x_2 + M_3x_3 - M*$


> $\sum_{j = 1}^3 w_2+_j  x_j ==> w_3x_1 + w_4x_2 + w_5x_3$







In [29]:
#Запишем ограничения задачи
w1, w2, w3, w4, w5 = sm.symbols('w1 w2 w3 w4 w5') #Наши лямбды
z1 = x1 + x2 + x3 - 1
z2 = M_avan_1 * x1 + M_mvid_2 * x2 + M_vjgz_3 * x3 - M_range
z3 = w3 * x1 + w4 * x2 + w5 * x3

print(f'Ограничения:\n'
f'{z1} = 0\n'
f'{z2} = 0\n'
f'{z3} = 0')

Ограничения:
x1 + x2 + x3 - 1 = 0
0.00112476301751057*x1 + 0.00604359189422394*x2 + 0.0031135893613839*x3 - 0.004 = 0
w3*x1 + w4*x2 + w5*x3 = 0


In [30]:
#Запишу функцию Лагранжа
L = Cx_x + w1 * z1 + w2 * z2 - z3

print(f'Функция Лагранжа выглядит следующим образом:\n{L}')

Функция Лагранжа выглядит следующим образом:
w1*(x1 + x2 + x3 - 1) + w2*(0.00112476301751057*x1 + 0.00604359189422394*x2 + 0.0031135893613839*x3 - 0.004) - w3*x1 - w4*x2 - w5*x3 + x1*(0.00141685843140227*x1 + 8.63479402469543e-5*x2 - 0.000249744668073191*x3) + x2*(8.63479402469543e-5*x1 + 0.00226351540928585*x2 + 0.00159762813078828*x3) + x3*(-0.000249744668073191*x1 + 0.00159762813078828*x2 + 0.00254293795517537*x3)


In [33]:
#Запишу условия стационарности
L_x1 = L.diff(x1)
L_x2 = L.diff(x2)
L_x3 = L.diff(x3)

print(f'Получившаяся система уравнений\n'
f'dL/dx1 = {L_x1} = 0\n'
f'dL/dx2 = {L_x2} = 0\n'
f'dL/dx3 = {L_x3} = 0\n'
f'{z1} = 0\n'
f'{z2} = 0\n'
f'{z3} = 0\n')

Получившаяся система уравнений
dL/dx1 = w1 + 0.00112476301751057*w2 - w3 + 0.00283371686280454*x1 + 0.000172695880493909*x2 - 0.000499489336146381*x3 = 0
dL/dx2 = w1 + 0.00604359189422394*w2 - w4 + 0.000172695880493909*x1 + 0.00452703081857169*x2 + 0.00319525626157656*x3 = 0
dL/dx3 = w1 + 0.0031135893613839*w2 - w5 - 0.000499489336146381*x1 + 0.00319525626157656*x2 + 0.00508587591035074*x3 = 0
x1 + x2 + x3 - 1 = 0
0.00112476301751057*x1 + 0.00604359189422394*x2 + 0.0031135893613839*x3 - 0.004 = 0
w3*x1 + w4*x2 + w5*x3 = 0



Для решения системы уравнений нужно рассмотреть 8 случаев:


1.   w3 = 0, w4 = 0, w5 = 0
2.   w2 = 0, w3 = 0, w4 = 0
3.   w1 = 0, w2 = 0, w3 = 0
4.   w2 = 0, w4 = 0, w5 = 0
5.   w2 = 0, w3 = 0, w5 = 0
6.   Новый пункт
7.   Новый пункт
8.   Новый пункт




**Воспользуюсь следующей формулой**

> $A_{m,n} * X_{m,1} = B_{m,1}$



In [32]:
#Запишу вектор для решения системы линейных уравнений
Ur_Vector = np.array([0.0, 0.0, 0.0, float(1), M_range])

#Случай 1: w3 = 0, w4 = 0, w5 = 0
Ur_Matrix_1 = np.array([[float(L_x1.diff(x1)), float(L_x1.diff(x2)), float(L_x1.diff(x3)), float(L_x1.diff(w1)), float(L_x1.diff(w2))],
                     [float(L_x2.diff(x1)), float(L_x2.diff(x2)), float(L_x2.diff(x3)), float(L_x2.diff(w1)), float(L_x2.diff(w2))],
                     [float(L_x3.diff(x1)), float(L_x3.diff(x2)), float(L_x3.diff(x3)), float(L_x3.diff(w1)), float(L_x3.diff(w2))],
                     [1.0, 1.0, 1.0, 0.0, 0.0], 
                     [float(z2.diff(x1)), float(z2.diff(x2)), float(z2.diff(x3)), float(z2.diff(w1)), float(z2.diff(w2))]])
#Случай 2: w2 = 0, w3 = 0, w4 = 0
Ur_Matrix_2 = np.array([[float(L_x1.diff(x1)), float(L_x1.diff(x2)), float(L_x1.diff(x3)), float(L_x1.diff(w1)), float(L_x1.diff(w5))],
                     [float(L_x2.diff(x1)), float(L_x2.diff(x2)), float(L_x2.diff(x3)), float(L_x2.diff(w1)), float(L_x2.diff(w5))],
                     [float(L_x3.diff(x1)), float(L_x3.diff(x2)), float(L_x3.diff(x3)), float(L_x3.diff(w1)), float(L_x3.diff(w5))],
                     [1.0, 1.0, 1.0, 0.0, 0.0], 
                     [float(z2.diff(x1)), float(z2.diff(x2)), float(z2.diff(x3)), float(z2.diff(w1)), float(z2.diff(w5))]])
#Случай 3: w1 = 0, w2 = 0, w3 = 0
Ur_Matrix_3 = np.array([[float(L_x1.diff(x1)), float(L_x1.diff(x2)), float(L_x1.diff(x3)), float(L_x1.diff(w4)), float(L_x1.diff(w5))],
                     [float(L_x2.diff(x1)), float(L_x2.diff(x2)), float(L_x2.diff(x3)), float(L_x2.diff(w4)), float(L_x2.diff(w5))],
                     [float(L_x3.diff(x1)), float(L_x3.diff(x2)), float(L_x3.diff(x3)), float(L_x3.diff(w4)), float(L_x3.diff(w5))],
                     [1.0, 1.0, 1.0, 0.0, 0.0], 
                     [float(z2.diff(x1)), float(z2.diff(x2)), float(z2.diff(x3)), float(z2.diff(w4)), float(z2.diff(w5))]])
#Случай 4: w2 = 0, w4 = 0, w5 = 0
Ur_Matrix_4 = np.array([[float(L_x1.diff(x1)), float(L_x1.diff(x2)), float(L_x1.diff(x3)), float(L_x1.diff(w1)), float(L_x1.diff(w3))],
                     [float(L_x2.diff(x1)), float(L_x2.diff(x2)), float(L_x2.diff(x3)), float(L_x2.diff(w1)), float(L_x2.diff(w3))],
                     [float(L_x3.diff(x1)), float(L_x3.diff(x2)), float(L_x3.diff(x3)), float(L_x3.diff(w1)), float(L_x3.diff(w3))],
                     [1.0, 1.0, 1.0, 0.0, 0.0], 
                     [float(z2.diff(x1)), float(z2.diff(x2)), float(z2.diff(x3)), float(z2.diff(w1)), float(z2.diff(w3))]])
#Случай 5: w2 = 0, w3 = 0, w5 = 0
Ur_Matrix_5 = np.array([[float(L_x1.diff(x1)), float(L_x1.diff(x2)), float(L_x1.diff(x3)), float(L_x1.diff(w1)), float(L_x1.diff(w4))],
                     [float(L_x2.diff(x1)), float(L_x2.diff(x2)), float(L_x2.diff(x3)), float(L_x2.diff(w1)), float(L_x2.diff(w4))],
                     [float(L_x3.diff(x1)), float(L_x3.diff(x2)), float(L_x3.diff(x3)), float(L_x3.diff(w1)), float(L_x3.diff(w4))],
                     [1.0, 1.0, 1.0, 0.0, 0.0], 
                     [float(z2.diff(x1)), float(z2.diff(x2)), float(z2.diff(x3)), float(z2.diff(w1)), float(z2.diff(w4))]])
#Случай 6: w1 = 0, w4 = 0, w5 = 0
Ur_Matrix_6 = np.array([[float(L_x1.diff(x1)), float(L_x1.diff(x2)), float(L_x1.diff(x3)), float(L_x1.diff(w2)), float(L_x1.diff(w3))],
                     [float(L_x2.diff(x1)), float(L_x2.diff(x2)), float(L_x2.diff(x3)), float(L_x2.diff(w2)), float(L_x2.diff(w3))],
                     [float(L_x3.diff(x1)), float(L_x3.diff(x2)), float(L_x3.diff(x3)), float(L_x3.diff(w2)), float(L_x3.diff(w3))],
                     [1.0, 1.0, 1.0, 0.0, 0.0], 
                     [float(z2.diff(x1)), float(z2.diff(x2)), float(z2.diff(x3)), float(z2.diff(w2)), float(z2.diff(w3))]])
#Случай 7: w1 = 0, w3 = 0, w5 = 0
Ur_Matrix_7 = np.array([[float(L_x1.diff(x1)), float(L_x1.diff(x2)), float(L_x1.diff(x3)), float(L_x1.diff(w2)), float(L_x1.diff(w4))],
                     [float(L_x2.diff(x1)), float(L_x2.diff(x2)), float(L_x2.diff(x3)), float(L_x2.diff(w2)), float(L_x2.diff(w4))],
                     [float(L_x3.diff(x1)), float(L_x3.diff(x2)), float(L_x3.diff(x3)), float(L_x3.diff(w2)), float(L_x3.diff(w4))],
                     [1.0, 1.0, 1.0, 0.0, 0.0], 
                     [float(z2.diff(x1)), float(z2.diff(x2)), float(z2.diff(x3)), float(z2.diff(w2)), float(z2.diff(w4))]])
#Случай 8: w1 = 0, w3 = 0, w4 = 0
Ur_Matrix_8 = np.array([[float(L_x1.diff(x1)), float(L_x1.diff(x2)), float(L_x1.diff(x3)), float(L_x1.diff(w2)), float(L_x1.diff(w5))],
                     [float(L_x2.diff(x1)), float(L_x2.diff(x2)), float(L_x2.diff(x3)), float(L_x2.diff(w2)), float(L_x2.diff(w5))],
                     [float(L_x3.diff(x1)), float(L_x3.diff(x2)), float(L_x3.diff(x3)), float(L_x3.diff(w2)), float(L_x3.diff(w5))],
                     [1.0, 1.0, 1.0, 0.0, 0.0], 
                     [float(z2.diff(x1)), float(z2.diff(x2)), float(z2.diff(x3)), float(z2.diff(w2)), float(z2.diff(w5))]])
#Случай 9: w1 = 0, w2 = 0, w5 = 0
Ur_Matrix_9 = np.array([[float(L_x1.diff(x1)), float(L_x1.diff(x2)), float(L_x1.diff(x3)), float(L_x1.diff(w3)), float(L_x1.diff(w4))],
                     [float(L_x2.diff(x1)), float(L_x2.diff(x2)), float(L_x2.diff(x3)), float(L_x2.diff(w3)), float(L_x2.diff(w4))],
                     [float(L_x3.diff(x1)), float(L_x3.diff(x2)), float(L_x3.diff(x3)), float(L_x3.diff(w3)), float(L_x3.diff(w4))],
                     [1.0, 1.0, 1.0, 0.0, 0.0], 
                     [float(z2.diff(x1)), float(z2.diff(x2)), float(z2.diff(x3)), float(z2.diff(w3)), float(z2.diff(w4))]])
#Случай 10: w1 = 0, w2 = 0, w4 = 0
Ur_Matrix_10 = np.array([[float(L_x1.diff(x1)), float(L_x1.diff(x2)), float(L_x1.diff(x3)), float(L_x1.diff(w3)), float(L_x1.diff(w5))],
                     [float(L_x2.diff(x1)), float(L_x2.diff(x2)), float(L_x2.diff(x3)), float(L_x2.diff(w3)), float(L_x2.diff(w5))],
                     [float(L_x3.diff(x1)), float(L_x3.diff(x2)), float(L_x3.diff(x3)), float(L_x3.diff(w3)), float(L_x3.diff(w5))],
                     [1.0, 1.0, 1.0, 0.0, 0.0], 
                     [float(z2.diff(x1)), float(z2.diff(x2)), float(z2.diff(x3)), float(z2.diff(w3)), float(z2.diff(w5))]])

answer1 = np.linalg.solve(Ur_Matrix_1, Ur_Vector)
answer2 = np.linalg.solve(Ur_Matrix_2, Ur_Vector)
answer3 = np.linalg.solve(Ur_Matrix_3, Ur_Vector)
answer4 = np.linalg.solve(Ur_Matrix_4, Ur_Vector)
answer5 = np.linalg.solve(Ur_Matrix_5, Ur_Vector)
answer6 = np.linalg.solve(Ur_Matrix_6, Ur_Vector)
answer7 = np.linalg.solve(Ur_Matrix_7, Ur_Vector)
answer8 = np.linalg.solve(Ur_Matrix_8, Ur_Vector)
answer9 = np.linalg.solve(Ur_Matrix_9, Ur_Vector)
answer10 = np.linalg.solve(Ur_Matrix_10, Ur_Vector)

print(answer1, answer2, answer3, answer4, answer5, answer6, answer7, answer8, answer9, answer10, sep='\n')



[ 0.39125996  0.5681083   0.04063174 -0.00082463 -0.32176497]
[ 0.65914981  0.7499465  -0.40909631 -0.0022017  -0.00221528]
[0.07814647 0.3555732  0.56628033 0.0034326  0.00397715]
[ 0.19278582  0.43338807  0.37382611 -0.00318972 -0.0027553 ]
[ 0.45130176  0.60886346 -0.06016522 -0.00141406  0.00122798]
[ 0.46046079  0.61508042 -0.07554121 -0.43395289  0.00096068]
[ 0.30726103  0.51109153  0.18164745 -0.77191655 -0.00171795]
[ 0.23084057  0.45921887  0.30994056 -0.51444617  0.00132656]
[ 0.6571727   0.74860448 -0.40577718  0.0021942   0.00220588]
[ 1.69838226  1.45535643 -2.15373868  0.00613984 -0.00715173]


In [43]:
print(answer1[0] + answer1[1] + answer1[2])
print(answer2[0] + answer2[1] + answer2[2])
print(answer3[0] + answer3[1] + answer3[2])
print(answer4[0] + answer4[1] + answer4[2])
print(answer5[0] + answer5[1] + answer5[2])
print(answer6[0] + answer6[1] + answer6[2])
print(answer7[0] + answer7[1] + answer7[2])
print(answer8[0] + answer8[1] + answer8[2])
print(answer9[0] + answer9[1] + answer9[2])
print(answer10[0] + answer10[1] + answer10[2])

1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0
1.0


In [27]:
#Случай 11:
Ur_Matrix_11 = np.array([[float(L_x1.diff(w1)), float(L_x1.diff(w3)), float(L_x1.diff(w4)), float(L_x1.diff(w5)), float(L_x1.diff(x1))],
                     [float(L_x2.diff(w1)), float(L_x2.diff(w3)), float(L_x2.diff(w4)), float(L_x2.diff(w5)), float(L_x2.diff(x1))],
                     [float(L_x3.diff(w1)), float(L_x3.diff(w3)), float(L_x3.diff(w4)), float(L_x3.diff(w5)), float(L_x3.diff(x1))],
                     [0.0, 0.0, 0.0, 0.0, 1.0], 
                     [float(z2.diff(w1)), float(z2.diff(w3)), float(z2.diff(w4)), float(z2.diff(w5)), float(z2.diff(x1))]])

answer11 = np.linalg.solve(Ur_Matrix_11, Ur_Vector)

print(answer11, sep='\n')

LinAlgError: ignored